### ISSUE TITLE
[k8s] Surface errors for GPU labelling script if `nvidia` is not set as container runtime

### ISSUE THREAD

**Issue by romilbhardwaj** [skypilot-org/skypilot#2958]
If `nvidia` is not set as the default runtime (e.g., in RKE2 if the user follows [default installation instructions](https://docs.nvidia.com/datacenter/cloud-native/gpu-operator/latest/getting-started.html#install-nvidia-gpu-operator), instead of [RKE2 instructions](https://docs.nvidia.com/datacenter/cloud-native/gpu-operator/latest/getting-started.html#custom-configuration-for-runtime-containerd)), `nvidia-smi` will fail inside the container. As a result:

1. GPU labelling script fails. But this error is not surfaced unless the user runs `kubectl logs <labelling pod> -n kube-system`.
2. If the user manually labels the pods, `nvidia-smi` won't work (even though other gpu checks work, e.g., `nvcc`).

If possible, we should detect and validate runtime during `sky check` (possibly an expensive call?) and consider making the GPU labelling script synchronous, where we wait for job to complete and surface errors, if any.

**Comment by chymian** [comment]
I just spend 3+ days to get a k3s cluster with GPU support `gpu-operator` up and running,
One thing I already recognized is, that Pods need to get the `runtimeClass` set to `nvidia` or `nvidia-experimental` to be able to access a GPUs

in the `gpu_test_pod.yaml`

```patch
   name: nvidia-smi
 spec:
+  runtimeClassName: nvidia
   restartPolicy: Never
```
```bash
kubectl get runtimeclasses.node.k8s.io 
NAME                  HANDLER               AGE
…
nvidia                nvidia                18h
nvidia-experimental   nvidia-experimental   36h
…
```

In this context:
new to skypilot and k8s (but not to linux), I only did this detour because I did read On-Prem is/will be deprecated?
given all the hassle to get the k3s up and running (on an external node with a zerotier-vpn and the `gpu-operator` for propper GPU-support, faceing years old bugs), IMHO using k8s as a ___easy and handy___ on-prem surrogate is not given.
can you please keep a simple on-prem/docker solution? thinking SOHO & SME market.

**Comment by romilbhardwaj** [comment]
Hi @chymian  - sorry to hear you had trouble setting up your k8s cluster. I would recommend using RKE2 over K3s, since K3s does not provide easy support to change the default container runtime and RKE2 is quite straight forward to setup.  Here's a [quick guide for installing RKE2](https://gist.github.com/romilbhardwaj/40e542ba09bb951307a9664f023aa410) I put together. LMK if it is helpful, happy to formalize it into our docs.

If you still plan on using k3s, #3019 will allow specifying a custom `runtimeClassName`, and after #3034 we should be able to autodetect k3s and apply `runtimeClassName`. Meanwhile, we will update the docs to have these pointers to suggest avoid using k3s for now. 

Are you on a single node cluster? If you want an easy way to try SkyPilot + K8s on a single node, consider trying `sky local up` on your GPU machine. This will create a local kubernetes cluster in a much simpler manner. Note that this cluster will be accessible only on the machine where it is run.

For now, we do not plan on supporting raw on-prem since handling diverse user environments becomes a challenge without any form of containerization, and orchestrating containers across nodes brings us to re-implementing kubernetes.

**Comment by chymian** [comment]
@romilbhardwaj 
rancher deployed cluster was where i was starting. didn't work due to the cattle-cluster-agent could not call back via private nameserver and VPN.  

I could only get k3s local modified setup and import into rancher to work.
with manual installed RKE2, only the cattle-cluster-agent does not use the configured DNS (coredns), somehow… :(
but that's just rancher. not really needed for a skypilot target cluster....

setting up RKE2 again manually and trying to get the gpu-operator to run also failed.

days of debugging to just get the underlaying infrastructur up and running, without being able to push one job cost too much time and effort. I gona wait for [#3019](https://github.com/skypilot-org/skypilot/pull/3019)  & [#3034](https://github.com/skypilot-org/skypilot/issues/3034)

and thanks for your quick guide. I would definitly like to see that in the docs.

**Comment by chymian** [comment]
hello @romilbhardwaj ,
I was just checking back on the k3s issue. 

since [#3019](https://github.com/skypilot-org/skypilot/pull/3019) got merged last week.  
is k3s  add `runtimeClassName: nvidia` auomatically supposed to work?

#### I tested it today with 1.0.0.dev20240204:  
the gpu-labeler finished it's job, but did not label anything (here 4 x rtx3070). 
and when I see it right, the only GPUs you support are:
```
    canonical_gpu_names = [
        'A100-80GB', 'A100', 'A10G', 'H100', 'K80', 'M60', 'T4g', 'T4', 'V100', 
        'A10', 'P100', 'P40', 'P4', 'L4', 'A6000'
 ``` 
 
 is that only true for the labeler or true for whole skypilot?  

 and can pls you extend to support non-DC GPUs? 
 like my 'old crypto RIG'  has 4 x RTX3070 and serves  between 2 and 4 LLMs flawlessly. 
 
 I did further testing by starting the examples/jupyter_lab.yaml
 > added `accelerators: RTX3070`
 
 ```bash
sky launch -c jupyter jupyter_lab.yaml
Task from YAML spec: jupyter_lab.yaml
I 02-05 17:52:25 optimizer.py:1222] No resource satisfying <Cloud>({'RTX3070': 1}, ports=['29324']) on [AWS, Kubernetes].
sky.exceptions.ResourcesUnavailableError: Catalog and kubernetes cluster does not contain any instances satisfying the request:
Task(run='jupyter lab --port 2...')
  resources: <Cloud>({'RTX3070': 1}, ports=['29324']).
```

and further gpu-checks:

```bash
$ sky show-gpus RTX3070
Resources 'RTX3070' not found. Try 'sky show-gpus --all' to show available accelerators.
```
```bash
sky show-gpus --cloud kubernetes
No GPUs found in Kubernetes cluster. If your cluster contains GPUs, make sure nvidia.com/gpu resource is available on the nodes and the node labels for identifying GPUs (e.g., skypilot.co/accelerators) are setup correctly. To further debug, run: sky check.
```
```bash
 kubectl describe node gulag|grep skypilot
                    skypilot.co/accelerators=rtx3070
```
label was manually added.

**Comment by romilbhardwaj** [comment]
Hey @chymian, thanks for checking in.

> is k3s add runtimeClassName: nvidia auomatically supposed to work?

No, not as of current master. #3148 will introduce this feature. Till then, you will need to add the following to your `~/.sky/config.yaml`:

```
kubernetes:
  pod_config:
    spec:
      runtimeClassName: nvidia
```

More docs [here](https://skypilot.readthedocs.io/en/latest/reference/config.html).

> is that only true for the labeler or true for whole skypilot?

That GPU list is only for canonical naming purposes, so that the GPU name detected aligns with the names we use in SkyPilot. If a GPU other than those in the list is detected, the name from nvidia-smi (`nvidia-smi --query-gpu=name --format=csv,noheader,nounits`) will be used to label. After #3148, the labelling script will work with K3s too and your rtx3070 should get labelled.

For your setup, what's the output of `sky check`? From your `sky show-gpus --cloud kubernetes` output, seems like `nvidia.com/gpu` resource might be missing. Can you confirm you see it in `kubectl describe node <..>`?

**Comment by romilbhardwaj** [comment]
Closed with #3148.
